functorch/csrc/BatchRulesScatterOps.cpp

```diff
@@ -111,6 +111,45 @@ Tensor index_plumbing(const Tensor & self, const List<optional<Tensor>> & indice
   return makeBatched(std::get<0>(results), std::get<1>(results), cur_level);
 }
 
+namespace {
+  std::tuple<Tensor, std::vector<optional<Tensor>>, Tensor>
+  index_put_batch_rule_helper(const Tensor &self,
+                              optional<int64_t> self_bdim,
+                              ArrayRef<optional<Tensor>> indices,
+                              ArrayRef<optional<int64_t>> indices_bdims,
+                              const Tensor &values,
+                              optional<int64_t> values_bdim) {
+
+    auto self_logical_rank = rankWithoutBatchDim(self, self_bdim);
+    auto values_logical_rank = rankWithoutBatchDim(values, values_bdim);
+    Tensor self_ = moveBatchDimToFront(self, self_bdim);
+    Tensor values_ = moveBatchDimToFront(values, values_bdim);
+    TORCH_INTERNAL_ASSERT(indices.size() == indices_bdims.size());
+    std::vector<optional<Tensor>> indices_ = batchIndices(indices, indices_bdims, self_.size(0), self_bdim, values_bdim);
+
+    // handle case when prefix unit dimensions are stripped in setitem.
+    // values of shape `[1, 1, 2]` becomes `[2]`
+    // This leads to problem if values was actually batched.
+    // so we re-append the unit dimensions after batch_dim.
```

Comment:
I'm really confused about why this happens, and I think the problem is it's unclear what the actual semantics of index_put_ are. I read your comment over at https://github.com/pytorch/functorch/issues/381#issuecomment-1017247434 -- do you have a sense of why it's OK for PyTorch to always change values of shape `[1, 1, 2]` to `[2]`?